### PR TITLE
Bk-CNN

### PR DESCRIPTION
**Pull request by evanyeohboonkhai**
Please check all file with the name started "CNN"

### CODE REVIEW

**Review by KianYang-Lee — COMMENTED** [review]
Sorry but kinda lost due to the sheer amount of scripts that you have included. Can you kindly specify a list of scripts that requires reviewing in this PR? Or another case is that you can standardize your naming of notebooks, either start with 04, or start with CNN, because now the naming is quite different. @evanyeohboonkhai , thank you.

**Comment by evanyeohboonkhai** [comment]
@KianYang-Lee  please have a look at these file .Thanks

04 - CNN- Univariate - Single-Step.ipynb

04 - CNN-Multivariate-Univariate-Mutistep.ipynb

04 - CNN-Multivariate-Univariate-Single-Step.ipynb

04 - CNN-Univariate-Multi-Step.ipynb

04 - Hybrid-CNN-LSTM Univariate.ipynb

File: notebook/Solution/04 - CNN-Univariate-Multi-Step.ipynb [hunk 1]
```diff
@@ -0,0 +1,1474 @@
...
+    "# Start Training \n",
+    "train_loss_shampoo,val_loss_shampoo = training(num_epochs_shampoo,train_iter_shampoo,test_iter_shampoo,optimiser_shampoo,loss_fn_shampoo,model_shampoo)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Evaluate\n",
+    "Plot the learning curve to evaluate the loss of each epoch "
```
Comment (KianYang-Lee) [inline notebook/Solution/04 - CNN-Univariate-Multi-Step.ipynb:1164]: This plot clearly shows overfitting, perhaps it will be helpful to either
1. Include a remark on this, or
2. Show how to revise the model to mitigate overfitting

File: notebook/Solution/04 - Hybrid-CNN-LSTM Univariate.ipynb [hunk 2]
```diff
@@ -0,0 +1,1292 @@
...
+      ]
+     },
+     "metadata": {
+      "needs_background": "light"
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "# Create Learining Curve \n",
```
Comment (KianYang-Lee) [inline notebook/Solution/04 - Hybrid-CNN-LSTM Univariate.ipynb:575]: 1. spelling error
2. The graph looks abit unusual for a loss curve, suggest to start from 1st epoch onwards to illustrate better?

**Review by KianYang-Lee — CHANGES_REQUESTED** [review]
Looked at the flow and logic of the codes, did not check the Markdown so will be suggesting based on flow and logic soundness, which were feedback in each file. General comments are:

1. A lot of repetitive parts across notebooks, it would be helpful to put them in a separate module and just call them as methods.
2.  Some showcase of basic datta analytics would be helpful to prime the viewers on how to perform effective data analysis, and also to prep them for the exam since we are going to check with them on that part.
3. Model performance wise does not look so satisfactory to me at the moment, could be due to the size of training dataset? Anyway would be great if the performance is better to showcase the power of DL.

Overall, a job well done. Hope that you can make some modifications based on the input and we will be ready to merge after that imho basing on the code logic.

File: notebook/Solution/04 - CNN-Multivariate-Univariate-Mutistep.ipynb [hunk 3]
```diff
@@ -0,0 +1,3026 @@
...
+      ]
+     },
+     "execution_count": 24,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "demand.insert(1,\"Day\",range(1,len(demand)+1))\n",
+    "demand\n",
```
[the remark this answers is not in the document]
Comment (evanyeohboonkhai) [reply notebook/Solution/04 - CNN-Multivariate-Univariate-Mutistep.ipynb:2188]: The cell is to show the student how they change index. Hence, 60 row is let them have a clear visualize of the data @KianYang-Lee

File: notebook/Solution/04 - CNN-Multivariate-Univariate-Mutistep.ipynb [hunk 4]
```diff
@@ -0,0 +1,3026 @@
...
+    "demand.set_index('Day')"
```
[the remark this answers is not in the document]
Comment (evanyeohboonkhai) [reply notebook/Solution/04 - CNN-Multivariate-Univariate-Mutistep.ipynb:2189]: The data index is updated @KianYang-Lee

File: notebook/Solution/04 - CNN-Multivariate-Univariate-Mutistep.ipynb [hunk 5]
```diff
@@ -0,0 +1,3026 @@
...
+       "4                210.517  "
+      ]
+     },
+     "execution_count": 25,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "demand_slice = demand.iloc[:,3::]\n",
```
[the remark this answers is not in the document]
Comment (evanyeohboonkhai) [reply notebook/Solution/04 - CNN-Multivariate-Univariate-Mutistep.ipynb:2342]: the demand_slice = demand.iloc[:,3::]  is to the column of data that we are interested in @KianYang-Lee

File: notebook/Solution/04 - CNN-Multivariate-Univariate-Mutistep.ipynb [hunk 6]
```diff
@@ -0,0 +1,3026 @@
...
...8A7PvBRCh2ghG3PJXn4T64eziEnOSe8YW8eh4y5NJEo76SnLOwKlSAe+7RPSB5TGD2mjZt9K3Mw3LPAw/oNCee/P9S6txKQofDoOfYj2j+SQMOha/k7pLjOTDmW667ogYDB8Lwrkls678Nx4sOtvXfxvCuSQwc6LmmuSIBlo+54Qb9PfPdd1a3ROTE4TDYcHYWFU/EU7lsRL7/fki7PlAsjYGfBEDPw0V+2JtC0HE7LRtEu94pKEivIDt1Cpo21b1aoCfCT5zo1kqy2xrGQ/AZJixYbk7DsytbVvdeuBgmPHHqDN+m/o8yR2LzTM8B0KHp5Vx+qge/hoxj+fqtBW+XLyfnNMOjj+qAcdIkq1uiVaoEV10l87DclZio3zuenuxut5P2599UfbIjHx7qTqm0q1nc5Tc+e6JPnqu5rSABlo8pXlyXw5J5WL7pixW/kRmxnQ41OxXo7++La0LY0WuZvfs9HI7AybOTkelgT7FlRON6kuk5DRvCU0/B8uVw5ozedsr9Wo2927eCjGJ8td6Dw4T//gu//nrJXU9MmoojfA9PNM+798rpk54vgiOI+z96oeBt8uXknGZo3BhattTDhL4y7yk2FlaudFmiaNQo6DihH8HDglFDFcHDguk4oZ8lCS0tV7y4vjiaNQuOHPHYaVamZZC2ZSP7whfSsfgYDr2R4rJMmS+QAMsHxcfrBRk7d1rdEnGx91Jmg6F4+tZbCnyMO+y9SS+9gfcX/GBiy6w1+4ff9fyr6Bj3/uCyy3SX//79+vczZ6BnT7fmFJUvHUaZ4y35/ZSHAqxOnXRP20XDhA6Hwcf/jKZ4agMG3+n+8HCTy6rRXPVnW8lkpi37rWBt8uXknGZ57DHYutV3UlHExengykUR8JWl+zFn73gyDR0QZhqZzNk7npWl+3mzlb6je3c99D9tmumH/nPnQWo8cRdzQxZR/pTBgrjvmfXMAIKDfDuE8e3WFVFSNsd3rT46i1JHW1A/umKBjzHqvrvhdElGLQmclA2f/5ACQPeYm9z7g6FDLx2CS093e05Rs/LxpJfewIat/+WjlW4qX15XX//iiwvaOOKL7zgduYFutZ7M93DE1EcGoU5H8shXzxasTdmTcyqlb90cUvUbzpQN77xjdUu0G2/U8/FczCuav28iXPwyUFnbi6LGjfWwqsnDhEM+nsMVY+uzK+IrSpW5E4
```
[the remark this answers is not in the document]
Comment (evanyeohboonkhai) [reply notebook/Solution/04 - CNN-Multivariate-Univariate-Mutistep.ipynb:2986]: which part do you? Can you attach the picture since it is hard to visualize base on the line in github

File: notebook/Solution/04 - CNN-Multivariate-Univariate-Mutistep.ipynb (hunk 5 above)
[the remark this answers is not in the document]
Comment (KianYang-Lee) [reply notebook/Solution/04 - CNN-Multivariate-Univariate-Mutistep.ipynb:2342]: why the 2nd colon though?

File: notebook/Solution/04 - CNN-Multivariate-Univariate-Mutistep.ipynb (hunk 6 above)
[the remark this answers is not in the document]
Comment (KianYang-Lee) [reply notebook/Solution/04 - CNN-Multivariate-Univariate-Mutistep.ipynb:2986]: You can try to run it on Jupyter Notebook and reference to the lines before it, that's the image thanks

File: notebook/Solution/04 - CNN-Multivariate-Univariate-Mutistep.ipynb (hunk 5 above)
[the remark this answers is not in the document]
Comment (evanyeohboonkhai) [reply notebook/Solution/04 - CNN-Multivariate-Univariate-Mutistep.ipynb:2342]: it means it sliced from the third column to the end of it @KianYang-Lee

[the remark this answers is not in the document]
Comment (KianYang-Lee) [reply notebook/Solution/04 - CNN-Multivariate-Univariate-Mutistep.ipynb:2342]: Are you sure 
`3::` = sliced from the third column to the end of it?

File: notebook/Solution/04 - CNN- Univariate - Single-Step.ipynb [hunk 7]
```diff
@@ -97,7 +97,7 @@
     {
```
Comment (KianYang-Lee) [inline notebook/Solution/04 - CNN- Univariate - Single-Step.ipynb:97]: Hyper Parameter should be "Hyperparameter" one word

File: notebook/Solution/04 - CNN-Multivariate-Univariate-Single-Step.ipynb [hunk 8]
```diff
@@ -2763,9 +1812,9 @@
    ],
    "source": [
     "# Zoom in by fill in start_epoch and end_epoch\n",
-    "start_epoch =90\n",
```
Comment (KianYang-Lee) [inline notebook/Solution/04 - CNN-Multivariate-Univariate-Single-Step.ipynb]: The losses cannot be further reduced?

**Review by KianYang-Lee — COMMENTED** [review]
Commented certain questions on relevant line. Overall looks good, thank you Boon Khai. Maybe take a look at the question first. Also, was just looking at the code logic and results without paying attention to the grammar and spelling of the notebook.

File: notebook/Solution/04 - CNN- Univariate - Single-Step.ipynb [hunk 9]
```diff
@@ -318,14 +309,17 @@
     "        \n",
     "        # For example Input:(N,C,Lin) Output : (N,C,Lout)\n",
     "        self.poo1 = nn.MaxPool1d(kernel_size=2)\n",
-    "        self.fc1 = nn.Linear(30, 10)\n",
+    "        \n",
+    "        self.conv2 = nn.Conv1d(in_channels = 30, out_channels = 20, kernel_size = 2)\n",
```
[the remark this answers is not in the document]
Comment (evanyeohboonkhai) [reply notebook/Solution/04 - CNN- Univariate - Single-Step.ipynb:313]: it can be any number, the kernel size must be smaller than the data sequence length. For example, if your data sequence 
length is 3, your kernel size should be 2. 
 If your kernel size is larger than the data sequence length, it will throw an error due to the data is too short for the kernel to stride, unless you use padding. However, padding will make decrease the accuracy (as your network goes deeper).
Hence, kernel size =2 is the most general way to use as I increase the layer of CNN (deeper the network)

[the remark this answers is not in the document]
Comment (KianYang-Lee) [reply notebook/Solution/04 - CNN- Univariate - Single-Step.ipynb:313]: Noted with your choice

**Review by KianYang-Lee — APPROVED** [review]

**Comment by KianYang-Lee** [comment]
Code logic looks good, merging it and deleting the branch.
